evm/utils/test_builder.py

```diff
@@ -0,0 +1,309 @@
+        self.info = OrderedDict([
+            ("comment", self.comment),
+            ("filledwith", self.filled_with),
+        ])
+        self.forks = forks
+
+        # pre state
+        self.env = merge(DEFAULT_ENV, env or OrderedDict())
+        self.pre_state = normalize_state(pre_state or OrderedDict())
+
+        # transaction
+        self.transaction_object = transaction
+        self.secret_key = secret_key
+
+        # post state
+        self.output = output
+        self.state_changes = normalize_state(state_changes)
+        self.logs = logs or []
+        self.log_hash = hash_log_entries(self.logs)
+        self.post_state = merge_states(self.pre_state, normalize_state(self.state_changes))
+        state_db = MainAccountStateDB(MemoryDB())
+        state_db.apply_state_dict(self.post_state)
+        self.post_state_root = state_db.root_hash
+
+    def to_dict(self):
+        return StateTestGroup([self]).to_dict()
+
+    def derive_variant(
+        self,
+        *,
+        name=None,
+        comment=None,
+        filled_with=None,
+        forks=None,
+        env=None,
+        pre_state=None,
+        transaction=None,
+        secret_key=None,
+        output=None,
+        state_changes=None,
+        logs=None
+    ):
+        kwargs = {
+            "name": name if name is not None else self.name,
+            "filled_with": filled_with if filled_with is not None else self.filled_with,
+            "forks": forks if forks is not None else self.forks,
+            "env": env if env is not None else self.env,
+            "pre_state": pre_state if pre_state is not None else self.pre_state,
+            "transaction": transaction if transaction is not None else self.transaction,
+            "secret_key": secret_key if secret_key is not None else self.secret_key,
+            "output": output if output is not None else self.output,
+            "state_changes": state_changes if state_changes is not None else self.state_changes,
+            "logs": logs if logs is not None else self.logs,
+        }
+        return self.__class__(**kwargs)
+
+
+class StateTestGroup:
+
+    def __init__(self, sub_tests, info=None):
+        self.sub_tests = sub_tests
+        self._info = info
+
+    @property
+    def name(self):
+        name = self.sub_tests[0].name
+        assert all(test.name == name for test in self.sub_tests)
+        return name
+
+    @property
+    def info(self):
+        if self._info is not None:
+            return self._info
+        else:
+            return self.sub_tests[0].info
+
+    @property
+    def env(self):
+        env = self.sub_tests[0].env
+        assert all(test.env == env for test in self.sub_tests)
+        return env
+
+    @property
+    def pre_state(self):
+        pre_state = self.sub_tests[0].pre_state
+        assert all(test.pre_state == pre_state for test in self.sub_tests)
+        return pre_state
+
+    def to_dict(self):
+        d = OrderedDict()
```

Comment:
`eth-utils` has an `eth_utils.to_ordered_dict` which will let you write this using the `yield` pattern (if you're inclined to).